source/windows/secure_channel_tls_handler.c

```diff
@@ -1264,131 +1264,92 @@ static int s_process_read_message(
-                        if (downstream_window > 0) {
-                            /* prevent a deadlock due to downstream handlers wanting more data, but we have an
-                               incomplete record, and the amount they're requesting is less than the size of a tls
-                               record. */
-                            size_t window_size = slot->window_size;
-                            if (!window_size &&
-                                aws_channel_slot_increment_read_window(slot, sc_handler->estimated_incomplete_size)) {
-                                err = AWS_OP_ERR;
-                            } else {
-                                sc_handler->estimated_incomplete_size = 0;
-                                err = AWS_OP_SUCCESS;
-                            }
-                        } else {
-                            /* Reset error to continue. Until all the message been read. */
-                            err = AWS_OP_SUCCESS;
-                        }
-                    }
-                    aws_byte_cursor_advance(&message_cursor, amount_to_move_to_buffer);
-                } else {
-                    /* Decrypt error, stop decrypting. */
+            /* prevent a deadlock due to downstream handlers wanting more data, but we have an incomplete
+                record, and the amount they're requesting is less than the size of a tls record. */
+            size_t downstream_window =
+                sc_handler->slot->adj_right ? aws_channel_slot_downstream_read_window(sc_handler->slot) : SIZE_MAX;
+            if (downstream_window > 0 && slot->window_size == 0) {
+                err = aws_channel_slot_increment_read_window(slot, sc_handler->estimated_incomplete_size);
+                if (err) {
                     break;
                 }
             }
+        } else {
+            /* we had a complete record. handle any left over extra data from the decrypt operation here. */
+            if (sc_handler->read_extra) {
+                size_t move_pos = sc_handler->buffered_read_in_data_buf.len - sc_handler->read_extra;
+                memmove(
+                    sc_handler->buffered_read_in_data_buf.buffer,
+                    sc_handler->buffered_read_in_data_buf.buffer + move_pos,
+                    sc_handler->read_extra);
+                sc_handler->buffered_read_in_data_buf.len = sc_handler->read_extra;
+                sc_handler->read_extra = 0;
+            } else {
+                sc_handler->buffered_read_in_data_buf.len = 0;
+            }
         }
-
-        if (!err) {
-            aws_mem_release(message->allocator, message);
-            return AWS_OP_SUCCESS;
-        }
-
-        aws_channel_shutdown(slot->channel, aws_last_error());
-        return AWS_OP_ERR;
     }
 
-    if (sc_handler->buffered_read_out_data_buf.len) {
-        if (s_process_pending_output_messages(handler)) {
-            return AWS_OP_ERR;
-        }
+    if (!err) {

```

Comment:
trivial: maybe do 
```
if(err) {
   // handle error
}
```
will be more straightforward.

Reply:
I was just keeping the old way, but whatever I've already changed so much